### PR TITLE
fix(lsp): properly handle border character list without highlight groups

### LINKED ISSUE #14691: vim.api.open_floating_preview inaccurate floating window position
HI, I switch to open_floating_preview recently.
I use this API to show signature help.

I found by default, the floating window is a bit too far away from cursor if I set ` relative = "cursor",     ` 

Screen Shot 2021-06-01 at 6 54 53 am

It might caused by `make_floating_popup_options`, so I dump the result of make_floating_popup_options:

```lua
{                                                                                                                                                                                  
  anchor = "SW",                                                                                                                                                                   
  border = { "╭", "─", "╮", "│", "╯", "─", "╰", "│" },                                                                                                                             
  col = 0,                                                                                                                                                                         
  height = 9,                                                                                                                                                                      
  relative = "cursor",                                                                                                                                                             
  row = -8,                                                                                                                                                                        
  style = "minimal",                                                                                                                                                               
  width = 109                                                                                                                                                                      
}     
```

What I would like is (previous behaviour)

float

Is there an additional setup I am missing when using open_floating_preview?

### PR DESCRIPTION
**Pull request by folke**
Fixes #14691

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -44,10 +44,20 @@ local function get_border_size(opts)
...
   else
-    height = height + vim.fn.strdisplaywidth(border[2][1])  -- top
-    height = height + vim.fn.strdisplaywidth(border[6][1])  -- bottom
-    width  = width  + vim.fn.strdisplaywidth(border[4][1])  -- right
-    width  = width  + vim.fn.strdisplaywidth(border[8][1])  -- left
+    local function border_width(id)
+      if type(border[id]) == "table" then
+        -- border specified as a table of <character, highlight group>
+        return vim.fn.strdisplaywidth(border[id][1])
+      elseif type(border[id]) == "table" then
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:51]: Am I missing something obvious here 😂

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:51]: right 🤦‍♂️
still waking up. just a sec

Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:51]: its all good :)

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: fixed in the meantime :)

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -44,10 +44,20 @@ local function get_border_size(opts)
...
-    width  = width  + vim.fn.strdisplaywidth(border[8][1])  -- left
+    local function border_width(id)
+      if type(border[id]) == "table" then
+        -- border specified as a table of <character, highlight group>
+        return vim.fn.strdisplaywidth(border[id][1])
+      elseif type(border[id]) == "string" then
+        -- border specified as a list of border characters
+        return vim.fn.strdisplaywidth(border[id])
+      end
+      error("floating preview border is not correct " .. vim.inspect(border))
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:55]: maybe more explicit error message pointing to the help docs and mentioning this should be a type or string. Also border can be "single", "double", "solid" but I think that's handled elsewhere

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: Good point, just added a reference to the docs. The single etc are handled in the code above indeed, so all good.

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -44,10 +44,20 @@ local function get_border_size(opts)
...
+      if type(border[id]) == "table" then
+        -- border specified as a table of <character, highlight group>
+        return vim.fn.strdisplaywidth(border[id][1])
+      elseif type(border[id]) == "string" then
+        -- border specified as a list of border characters
+        return vim.fn.strdisplaywidth(border[id])
+      end
+      error("floating preview border is not correct. Please refer to the docs |vim.api.nvim_open_win()|" .. vim.inspect(border))
+    end
+    height = height + border_width(2)  -- top
```
Comment (bfredl) [inline runtime/lua/vim/lsp/util.lua:57]: `strdisplaywidth` shouldn't be used for height, just whether there is a char (1) or not (0).

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:57]: Good point. I actually wonder, can you have a top border with something like `#\n#` that would be two then?

File: runtime/lua/vim/lsp/util.lua (hunk 2 above)
Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: fixed

File: runtime/lua/vim/lsp/util.lua (hunk 3 above)
Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: fixed

**Comment by mjlbach** [comment]
@bfredl this should be g2g since the height fix right?

**Comment by bfredl** [comment]
Yes.

**Pull request merged**